test/externalTests/euler.sh

```diff
@@ -44,7 +44,7 @@ function euler_test
         "${compile_only_presets[@]}"
         #ir-no-optimize           # Compilation fails with "YulException: Variable var_utilisation_307 is 6 slot(s) too deep inside the stack."
         #ir-optimize-evm-only     # Compilation fails with "YulException: Variable var_utilisation_307 is 6 slot(s) too deep inside the stack."
-        #ir-optimize-evm+yul      # Compilation fails with "YulException: Variable var_status_mpos is 3 too deep in the stack"
+        ir-optimize-evm+yul      # Compilation fails with "YulException: Variable var_status_mpos is 3 too deep in the stack"
```

Comment:
Looks like it compiles, but fails its tests: https://app.circleci.com/pipelines/github/ethereum/solidity/21737/workflows/3588ee0e-de5a-4b29-81ce-fe17cff72fbb/jobs/952228

Might be worth looking into. Could either be unrelated, or they have stack-to-memory incompatible inline assembly or a proper bug in the code transform.

Reply:
@cameel You said something about euler tests breaking - but it's probably not what I'm seeing here, right? This looks like it fails specifically for the ``ir-optimize-evm+yul`` run, right?

Reply:
No, the one I saw was completely different: https://app.circleci.com/pipelines/github/ethereum/solidity/21732/workflows/4b8b15e2-8d05-4a00-be6e-bd449ebf01dc/jobs/952012

```
Error: Timeout of 20000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/tmp/ext-test-Euler-7owwrh/ext/test/activateMarket.js)
```
```
Error: the string "expected error \"e/no-uniswap-pool-avail\" but no error was thrown" was thrown, throw an Error :)
```
```
Error: call revert exception (method="getPool(address,address,uint24)", errorArgs=null, errorName=null, errorSignature=null, reason=null, code=CALL_EXCEPTION, version=abi/5.5.0)
```

It was also a bunch of different errors and only a handful of tests failed. It went away after rerunning. Yours seem to be breaking tons of tests, all with the same error.

Reply:
> This looks like it fails specifically for the `ir-optimize-evm+yul` run, right?

Yeah, it's only that one. I'm looking into it.

From the stack trace looks like the revert is coming from [`revertBytes()`](https://github.com/euler-xyz/euler-contracts/blob/70cfa81ef67a76c44ac1e28249cadecd34f00a13/contracts/Base.sol#L87-L92). This is just a helper through. The relevant code seems to be [this part in `getNewMarketParameters()`](https://github.com/euler-xyz/euler-contracts/blob/70cfa81ef67a76c44ac1e28249cadecd34f00a13/contracts/modules/RiskManager.sol#L71-L93) instead even though I don't see it directly listed in the trace (maybe line numbers are off?). It's the only place that throws that error message.

Now the question is why it reverts. To figure it out we'll likely need to extract this part of code and play with it. Unfortunately it looks like it has external calls to other deployed contracts so it might be a bit inconvenient.

Reply:
Yeah... the assembly in here
https://github.com/euler-xyz/euler-contracts/blob/70cfa81ef67a76c44ac1e28249cadecd34f00a13/contracts/Euler.sol#L49
and here
https://github.com/euler-xyz/euler-contracts/blob/70cfa81ef67a76c44ac1e28249cadecd34f00a13/contracts/Proxy.sol#L18
are actually stack-to-memory incompatible - so setting it to forced-enabled is random luck, maybe it's that.

Reply:
And given that, it may not be worth it to investigate further for the time being.

Reply:
I mean - that's why this stuff doesn't "just work" right now - we disable the stack-to-memory stuff as soon as we cannot be sure that people don't just write to memory arbitrarily - currently rather crude by globally disabling it once there is any inline assembly anywhere whatsoever.

The plan is to allow people to annotate assembly as "memory safe" and then still do it - but we also need to be more fine-grained, i.e. *some* unsafe inline assembly in *some* paths should not prevent us from doing stack-to-memory anywhere else.

In this PR I just disabled all such safeguards to see what would happen - and for the others it seems fine - here it *might* be that the two snippets actually cause problems, but it may also be something completely unrelated...

The cases both seem proxy-related - in those cases people probably often avoid and want to avoid proper allocations to avoid the overhead...

Reply:
Scrolling through it again, it looks to me like all errors have the Proxy fallback and ``Euler.dispatch`` in their trace and those exactly contain such unsafe inline assembly - so it's probably that. FYI there's only a problem if a stack error occurs *while* you're writing to unsafe memory locations - it's a bit surprising that this happens in these cases, but apparently it does...

Reply:
Ah, right. In that case I think we really need to report it better regardless of how/if it gets fixed properly. If it fails due to "stack to deep" **and** stack-to-memory was disabled due to inline assembly, there should be a hint for the user that this might be the cause. Maybe we actually should have an INFO message every time this stuff gets disabled, regardless of "stack-too-deep".

Reply:
Yep - although there's not much point in that until we support annotating inline assembly as "memory safe" (until we have that, we could just suggest in bulk not to use inline assembly at all :-)).
But that we do need soon.

What I'd also actually be curious about is the gas differences between ``ir-optimize-evm+yul`` amd ``legacy-optimize-evm+yul`` on all the ones that work both ways in this PR... but it can wait until we have the benchmark PR merged.

Reply:
> FYI there's only a problem if a stack error occurs while you're writing to unsafe memory locations - it's a bit surprising that this happens in these cases, but apparently it does...

Shows that we really do need more of these external projects. Some important things surfaced this week (this and the hole in IR coverage for array copying) so I'm wondering if it might actually be worthwhile to add even more of them. I was going to stop after I finish the ones I have half-working now (ElementFi, PRB Math, Gnosis GP2) and maybe Rarible and Brink if I can get them to work at all. Maybe I should continue instead?

Reply:
I said it's surprising, since I didn't think these particular projects were likely to do something like this, but generally it's not surprising at all that cases like that occur :-). So it's really only the array copy thing.

So I think it's fine to stop after the half-working ones for the time being at least :-). The more of them it is the more likely it becomes that some will break, so maybe better to monitor the ones we have now or which will be added shortly for a while.